pkg/collectors/log_follower.go

```diff
@@ -0,0 +1,329 @@
+// the lines are then fed into a channel, in another gorotine they are de-duplicated and written to an output file.
+//
+// Overlap:
+// cmd       followDuration
+// |---------------|
+// since          cmd        followDuration
+// |---------------|---------------|
+// .............. since           cmd        followDuration
+// ..............  |---------------|---------------|
+//
+// This was done because oc logs and the kubeapi endpoint which it uses does not look back
+// over a log rotation event, nor does it continue to follow.
+//
+// Log aggregators would be preferred over this method however this requires extra infra
+// which might not be present in the environment.
+type LogsCollector struct {
+	lastPoll           time.Time
+	quit               chan os.Signal
+	lines              chan *ProcessedLine
+	seenLines          map[time.Time][]string
+	client             *clients.Clientset
+	logsOutputFileName string
+	wg                 sync.WaitGroup
+	pollInterval       int
+	lpLock             sync.RWMutex
+	withTimeStamps     bool
+	running            bool
+	pruned             bool
+}
+
+const (
+	LogsCollectorName = "Logs"
+	LogsInfo          = "log-line"
+)
+
+func (logs *LogsCollector) GetPollInterval() int {
+	return logs.pollInterval
+}
+
+func (logs *LogsCollector) IsAnnouncer() bool {
+	return false
+}
+
+func (logs *LogsCollector) SetLastPoll(pollTime time.Time) {
+	logs.lpLock.Lock()
+	logs.lastPoll = pollTime
+	logs.lpLock.Unlock()
+}
+
+// Start sets up the collector so it is ready to be polled
+func (logs *LogsCollector) Start() error {
+	go logs.writeToLogFile()
+	logs.running = true
+	return nil
+}
+
+func (logs *LogsCollector) consumeLine(line *ProcessedLine, writer io.StringWriter) {
+	if seen, ok := logs.seenLines[line.Timestamp]; ok {
+		for _, seenContent := range seen {
+			if seenContent == line.Content {
+				// Already have this line skip it
+				return
+			}
+		}
+	}
+	logs.seenLines[line.Timestamp] = append(logs.seenLines[line.Timestamp], line.Content)
+	logs.pruned = false
+	var err error
+	if logs.withTimeStamps {
+		_, err = writer.WriteString(line.Raw + "\n")
+	} else {
+		_, err = writer.WriteString(line.Content + "\n")
+	}
+	if err != nil {
+		log.Error(fmt.Errorf("failed to write log output to file"))
+	}
+}
+
+func (logs *LogsCollector) writeToLogFile() {
+	logs.wg.Add(1)
+	defer logs.wg.Done()
+
+	fileHandle, err := os.OpenFile(logs.logsOutputFileName, os.O_CREATE|os.O_WRONLY, logFilePermissions)
+	utils.IfErrorExitOrPanic(err)
+	defer fileHandle.Close()
+
+	for {
+		select {
+		case <-logs.quit:
+			// Consume the rest of the lines so we don't miss lines
+			for len(logs.lines) > 0 {
+				line := <-logs.lines
+				logs.consumeLine(line, fileHandle)
+			}
+			return
+		case line := <-logs.lines:
+			logs.consumeLine(line, fileHandle)
+		default:
+			if !logs.pruned {
+				log.Debug("logs: pruning seen lines")
+				logs.lpLock.RLock()
+				keepTime := logs.lastPoll.Add(-2 * followDuration)
```

Comment:
I think `keepTime` should be quite a bit longer; 15-30 mins or so. That should only be 5-10MB of logs, and removes any risk of weird behaviours if a short connection outage (but greater than 2x`followDuration`) happens and `sinceSeconds` ends up longer than `keepTime`